controllers/hvpa_controller.go

```diff
@@ -280,47 +274,67 @@ func (r *HvpaReconciler) deleteHvpaFinalizers(hvpa *autoscalingv1alpha1.Hvpa) {
 	}
 }
 
-func getVpaFromHvpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.VerticalPodAutoscaler, error) {
-	metadata := hvpa.Spec.Vpa.Template.ObjectMeta.DeepCopy()
-
-	labels := metadata.GetLabels()
-	if labels == nil || len(labels) == 0 {
-		// TODO: Could be done better as part of validation
-		return nil, fmt.Errorf("Need labels in VPA template")
+func (r *HvpaReconciler) reconcileVpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.VerticalPodAutoscalerStatus, error) {
+	selector, err := metav1.LabelSelectorAsSelector(hvpa.Spec.Vpa.Selector)
+	if err != nil {
+		log.Error(err, "Error converting vpa selector to selector")
+		return nil, err
 	}
 
-	if ownerRef := metadata.GetOwnerReferences(); len(ownerRef) != 0 {
-		// TODO: Could be done better as part of validation
-		return nil, fmt.Errorf("vpa template in hvpa object already has an owner reference")
+	// list all vpas to include the vpas that don't match the hvpa`s selector
+	// anymore but has the stale controller ref.
+	vpas := &vpa_api.VerticalPodAutoscalerList{}
+	err = r.List(context.TODO(), vpas, client.InNamespace(hvpa.Namespace))
+	if err != nil {
+		log.Error(err, "Error listing vpas")
+		return nil, err
 	}
 
-	if generateName := metadata.GetGenerateName(); len(generateName) != 0 {
-		log.V(3).Info("Warning", "Generate name provided in the vpa template will be ignored", generateName)
+	updatePolicy := hvpa.Spec.Vpa.UpdatePolicy.UpdateMode
+
+	// NOTE: filteredVpas are pointing to deepcopies of the cache, but this could change in the future.
+	// Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.2/pkg/cache/internal/cache_reader.go#L74
+	// if you need to modify them, you need to copy it first.
+	filteredVpas, err := r.claimVpas(hvpa, selector, vpas)
```

Comment:
Can we filtering here by filtering during the `List` call [above](https://github.com/gardener/hvpa-controller/pull/34/files#diff-6354ac486a904d10093c497f09f83a3eR287)?